### PR TITLE
Expose shard field stats

### PR DESCRIPTION
**Pull request by dnhatn**
Previously, we returned the number of segments and the total number of fields in those segments in NodeMappingStats (see #111123). However, the total number of fields returned in that PR might be very inaccurate for indices having large mappings but only a small number of actual fields.

This change returns a more accurate total number of fields using the Lucene FieldInfos from those segments. Since we need to acquire a searcher to compute this stats, we opt to compute it after a shard is refreshed and cache the result.

I will try to revert the change introduced in #111123.

Relates #111123

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 1]
```diff
@@ -4060,6 +4063,38 @@ public void afterRefresh(boolean didRefresh) {
...
+    private class RefreshShardFieldStatsListener implements ReferenceManager.RefreshListener {
+        @Override
+        public void beforeRefresh() {
+
+        }
+
+        @Override
+        public void afterRefresh(boolean didRefresh) {
+            if (shardFieldStats == null || didRefresh) {
+                try (var searcher = getEngine().acquireSearcher("shard_field_stats", Engine.SearcherScope.INTERNAL, Function.identity())) {
```
Comment (pxsalehi) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:4083]: nit: there seem to be a `acquireSearcher(String source, SearcherScope scope)` that could be used here.

File: server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java [hunk 2]
```diff
@@ -1784,6 +1785,56 @@ public void testExternalRefreshMetric() throws IOException {
...
+                    "f3": "foobar"
+                }
+                """);
+        }
+        shard.refresh("test");
+        stats = shard.getShardFieldStats();
+        assertThat(stats.numSegments(), equalTo(2));
+        // 9 + _id, _source, _version, _primary_term, _seq_no, f1, f1.keyword, f2, f2.keyword, f3, f3.keyword
+        assertThat(stats.totalFields(), equalTo(21));
+        closeShards(shard);
```
Comment (pxsalehi) [inline server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:1835]: nit: would it worth also doing a force merge and validate the stats result again?

**Review by pxsalehi — APPROVED** [review]
LGTM. Thanks for separating this out.

**Review by andreidan — APPROVED** [review]
LGTM thanks for adding this Nhat

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: Yes, fixed in https://github.com/elastic/elasticsearch/pull/111525/commits/0ee2b70bc4424da3938bb8f0c55e4b7ba3589912.

File: server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java (hunk 2 above)
Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:1835]: ++, added that assertion in https://github.com/elastic/elasticsearch/pull/111525/commits/a9109753f3d31ab3ba74fd4808b33c6c4b0bcf06

**Comment by dnhatn** [comment]
@pxsalehi @andreidan Thanks for reviewing.

**Pull request merged**